File: client/src/app/lib/transaction.ts
import { fetchdata } from "@/utils/Fetch";



const createTransaction = async (
    transaction_type : string,
    category: string,
    description: string,
    amount: number,
    transaction_date: number,
    transaction_month: number,
    transaction_year: number,
) => {
    try{
        const res = await fetchdata.post('/transaction/create', {
            transaction_type: transaction_type,
            category: category,
            description: description,
            amount: amount,
            transaction_date: transaction_date,
            transaction_month: transaction_month,
            transaction_year: transaction_year,
        });
        return res.data;
    }
    catch(error){
        console.error('Error creating transaction:', error);
        throw error;
    }
}

const getTransactionsByMonthYear = async (month: number, year: number) => {
    try{
        const res = await fetchdata.get(`/transaction/month/${month}/year/${year}`);
        return res.data;
    }
    catch(error){
        console.error('Error getting transactions:', error);
        throw error;
    }
}

const getTransactionsByDayMonthYear = async (day:number, month: number, year: number) => {
    try{
        const res = await fetchdata.get(`/transaction/day/${day}/month/${month}/year/${year}`);
        return res.data;
    } catch(error){
        console.error('Error getting transactions:', error);
        throw error;
    }
}
    
export {    
        createTransaction,
        getTransactionsByMonthYear ,
        getTransactionsByDayMonthYear,
    };

File: server/transaction/transaction.go
package transaction

import (
	"time"

	"gorm.io/gorm"
)

type Transaction struct {
	gorm.Model
	TransactionType  string  `gorm:"not null"` // Revenue, Expense, Other
	Category         string  `gorm:"not null"`
	Description      string  // Food, Travel, Utility
	Amount           float64 `gorm:"not null"` // Amount
	TransactionDate  int     `gorm:"not null"` // Date
	TransactionMonth int     `gorm:"not null"` // Month
	TransactionYear  int     `gorm:"not null"` // Year
}

type NewTransactionRequest struct {
	TransactionType  string  `json:"transaction_type"`
	Category         string  `json:"category"`
	Description      string  `json:"description"`
	Amount           float64 `json:"amount"`
	TransactionDate  int     `json:"transaction_date"`  // Date
	TransactionMonth int     `json:"transaction_month"` // Month
	TransactionYear  int     `json:"transaction_year"`  // Year
}

type TransactionResponse struct {
	CreateAt         time.Time `json:"create_at"`
	TransactionType  string    `json:"transaction_type"`
	Category         string    `json:"category"`
	Description      string    `json:"description"`
	Amount           float64   `json:"amount"`
	TransactionDate  int       `json:"transaction_date"`  // Date
	TransactionMonth int       `json:"transaction_month"` // Month
	TransactionYear  int       `json:"transaction_year"`  // Year
}

type TransactionSummaryResponse struct {
	TotalRevenue   float64                              `json:"total_revenue"`
	TotalExpense   float64                              `json:"total_expense"`
	TotalCredit    float64                              `json:"total_credit"`
	TotalRemaining float64                              `json:"total_remaining"`
	Transactions   []TransactionSummaryCategoryResponse `json:"transactions"`
}

type TransactionSummaryCategoryResponse struct {
	TransactionType string  `json:"transaction_type"`
	Category        string  `json:"category"`
	Amount          float64 `json:"amount"`
}

type SummaryRevenueExpenseResponse struct {
	TotalRevenue []float64 `json:"total_revenue"`
	TotalExpense []float64 `json:"total_expense"`
}
